docs/docs/reference/smart_contract_reference/testing/index.md

```diff
@@ -0,0 +1,120 @@
+---
+title: Testing
+---
+
+Aztec contracts can be tested in a variety of ways depending on the needs of a particular application and the complexity of the interactions they must support. 
+
+## Pure Noir tests
+
+Noir supports the `#[test]` annotation which can be used to write simple logic tests on isolated utility functions. These tests only make assertions on algorithms and cannot interact with protocol-specific constructs such as `storage` or `context`, but are extremely fast and can be useful in certain scenarios. 
+
+#include_code pure_noir_testing /noir-projects/noir-contracts/contracts/card_game_contract/src/cards.nr rust
+
+To learn more about Noir testing, please refer to [the docs](https://Noir-lang.org/docs/tooling/testing/)
+
+## TXE
+
+In order to interact with the protocol, aztec contracts leverage the power of oracles: functions that reach out to the outside world and are able to query and manipulate data outside of itself. The values returned by oracles are then constrained inside Noir and the modifications to the blockchain state later verified to adhere to the protocol rules by our kernel circuits.
+
+However, all of this is often not necessary to ensure the contract logic itself is sound, and all that we need is an entity to provide values consistent with real execution. This is where our TXE (Testing eXecution Environment) comes in!
+
+TXE is a JSON RPC server much like PXE, but provides an extra set of oracle functions called `cheatcodes` that allow developers to manipulate the state of the chain and simulate contract execution. Since TXE skips most of the checks, blockbuilding and other intrincacies of the aztec protocol, it is much faster to run than simulating everything in the sandbox.
+
+### Running TXE
+
+In order to use the TXE, it must be running on a known address. Assuming the default `http://localhost:8080`, contract tests would be run with:
+
+ `nargo test --oracle-resolver http://localhost:8080`
+
+:::warning
+Since TXE tests are written in Noir and executed with `nargo`, they all run in parallel. This also means every test creates their own isolated environment, so state modifications are local to each one of them.
+Furthermore, executing many tests in parallel might slow processing of the RPC calls down to the point of making them timeout. To control this timeout the `NARGO_FOREIGN_CALL_TIMEOUT` env variable is used.
+:::
+
+ ### Writing TXE contract tests in Noir
+
+`aztec-nr` provides an utility class called `TestEnvironment`, that should take care of the most common operations needed to setup contract testing. Setting up a new test environment with `TestEnvironment::new()` **will reset the current test's TXE state**
+
+#include_code txe_test_increment /noir-projects/noir-contracts/contracts/counter_contract/src/main.nr rust
+
+#### Deploying contracts
+
+```rust
```

Comment:
can these not be hardcoded?

Reply:
I've tried to provide examples wherever it made sense, but in some places they're just generic examples of usage. Don't really know what's best, too specific might be bad too.